<!DOCTYPE html>
<html lang="en"><head>
<script src="slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="slides_files/libs/quarto-html/popper.min.js"></script>
<script src="slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="slides_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="slides_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.1.175">

  <meta name="author" content="Carlos Ronchi">
  <title>An (opinionated) introduction to R for biologists</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="center">
  <h1 class="title">An (opinionated) introduction to R for biologists</h1>
  <p class="author">Carlos Ronchi</p>
</section>

<section>
<section id="introducing-quarto" class="title-slide slide level1 center">
<h1>Introducing Quarto</h1>

</section>
<section id="what-is-quarto" class="slide level2">
<h2>What is Quarto?</h2>
<blockquote>
<p>Quarto is a multi-language, next-generation version of R Markdown from RStudio, and includes dozens of new features and capabilities while at the same being able to render most existing Rmd files without modification.</p>
</blockquote>
<blockquote>
<p>R Markdown is a way to run R and generate reports in a “reproducible” manner</p>
</blockquote>
</section>
<section id="why-quarto" class="slide level2">
<h2>Why Quarto?</h2>
<ul>
<li>Whole ecosystem of best practices</li>
<li>Lots of support from their forums and stack exchange</li>
<li>Good tools that help you make better science</li>
</ul>
</section>
<section id="chunks" class="slide level2">
<h2>Chunks</h2>

<img data-src="images/code-1-options.png" class="r-stretch"></section></section>
<section id="projects" class="title-slide slide level1 center">
<h1>Projects</h1>
<p>A project is a handy way to go back and forth among your analysis. To create a new project in the RStudio IDE, use the Create Project command (button on top right corner with a R inside a cube).</p>

<img data-src="images/projects_new.png" class="r-stretch"></section>

<section>
<section id="folder-structure" class="title-slide slide level1 center">
<h1>Folder structure</h1>
<p>After creating the project, I always recommend to follow a structure of the folder. In the top folder, it is where the project file will be sitting, then there will be three different folder: - scripts - data - results</p>
</section>
<section id="scripts" class="slide level2">
<h2>Scripts</h2>
<p>This is where all your analysis scripts will be stored</p>
</section>
<section id="data" class="slide level2">
<h2>Data</h2>
<p>This is where all your original data is stored</p>
</section>
<section id="results" class="slide level2">
<h2>Results</h2>
<p>This is where all the intermediate results, images generated and reports will be stored.</p>
</section></section>
<section>
<section id="analysing-qpcr" class="title-slide slide level1 center">
<h1>Analysing qPCR</h1>

</section>
<section id="initial-check" class="slide level2">
<h2>Initial check</h2>
<p>The first thing we need to do is to check the excel file. Usually there are some missing rows or headers that makes our life difficult. We want to end up with an excel file that has the following sheet:</p>

<img src="images/excel_template.png" alt="Excel template" class="r-stretch"></section>
<section id="dataframes" class="slide level2">
<h2>Dataframes</h2>
<p>Nomenclature used in R to define tables. They are the most important object here, all the operations are performed in a dataframe. Example of a dataframe in R:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  Sepal.Length Sepal.Width Petal.Length Petal.Width Species
1          5.1         3.5          1.4         0.2  setosa
2          4.9         3.0          1.4         0.2  setosa
3          4.7         3.2          1.3         0.2  setosa
4          4.6         3.1          1.5         0.2  setosa
5          5.0         3.6          1.4         0.2  setosa
6          5.4         3.9          1.7         0.4  setosa</code></pre>
</div>
</div>
</section>
<section id="loading" class="slide level2">
<h2>Loading</h2>
<p>We start by loading the data. For this we can use the package <code>readxl</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2" data-code-line-numbers="|1|4|5|6|7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb2-2"><a href="#cb2-2"></a></span>
<span id="cb2-3"><a href="#cb2-3"></a>qpcr <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(</span>
<span id="cb2-4"><a href="#cb2-4"></a>    <span class="st">"../data/qPCR PA and ML EMT Cells ZEB2 SNAI1 TWIST1 .xls"</span>, <span class="co"># path to file</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>    <span class="at">sheet =</span> <span class="st">"Results"</span>,</span>
<span id="cb2-6"><a href="#cb2-6"></a>    <span class="at">range =</span> <span class="st">"A40:I220"</span>, </span>
<span id="cb2-7"><a href="#cb2-7"></a>    <span class="at">col_names =</span> <span class="cn">TRUE</span> <span class="co"># should the first row correspond to column names?</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="inspecting-the-data" class="slide level2">
<h2>Inspecting the data</h2>
<p><code>head</code> shows the first 5 rows of your dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">head</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
   Well `Well Position` Omit  `Sample Name` Target…¹ Task  Repor…² Quenc…³ CT   
  &lt;dbl&gt; &lt;chr&gt;           &lt;lgl&gt; &lt;chr&gt;         &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;
1   195 I3              FALSE 1. eGFP N=1   CDH1     UNKN… SYBR    None    18.4…
2   196 I4              FALSE 1. eGFP N=1   CDH1     UNKN… SYBR    None    18.2…
3   219 J3              FALSE 1. eGFP N=1   CDH1     UNKN… SYBR    None    18.2…
4   245 K5              FALSE 10. ZEB2 N=3  CDH1     UNKN… SYBR    None    18.5…
5   246 K6              FALSE 10. ZEB2 N=3  CDH1     UNKN… SYBR    None    18.4…
6   269 L5              FALSE 10. ZEB2 N=3  CDH1     UNKN… SYBR    None    18.4…
# … with abbreviated variable names ¹​`Target Name`, ²​Reporter, ³​Quencher</code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>The function <code>glimpse</code> shows in each row what can be found in each column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="fu">glimpse</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 180
Columns: 9
$ Well            &lt;dbl&gt; 195, 196, 219, 245, 246, 269, 293, 294, 317, 341, 342,…
$ `Well Position` &lt;chr&gt; "I3", "I4", "J3", "K5", "K6", "L5", "M5", "M6", "N5", …
$ Omit            &lt;lgl&gt; FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE…
$ `Sample Name`   &lt;chr&gt; "1. eGFP N=1", "1. eGFP N=1", "1. eGFP N=1", "10. ZEB2…
$ `Target Name`   &lt;chr&gt; "CDH1", "CDH1", "CDH1", "CDH1", "CDH1", "CDH1", "CDH1"…
$ Task            &lt;chr&gt; "UNKNOWN", "UNKNOWN", "UNKNOWN", "UNKNOWN", "UNKNOWN",…
$ Reporter        &lt;chr&gt; "SYBR", "SYBR", "SYBR", "SYBR", "SYBR", "SYBR", "SYBR"…
$ Quencher        &lt;chr&gt; "None", "None", "None", "None", "None", "None", "None"…
$ CT              &lt;chr&gt; "18.429319381713867", "18.256540298461914", "18.203689…</code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>The columns Task, Reporter and Quencher are not really necessary here for the following analysis, so we can drop it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>columns_to_drop <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Task"</span>, <span class="st">"Reporter"</span>, <span class="st">"Quencher"</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a>qpcr <span class="ot">&lt;-</span> qpcr <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>dplyr<span class="sc">::</span><span class="fu">all_of</span>(columns_to_drop))</span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="fu">colnames</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Well"          "Well Position" "Omit"          "Sample Name"  
[5] "Target Name"   "CT"           </code></pre>
</div>
</div>
</section>
<section id="cleaning" class="slide level2">
<h2>Cleaning</h2>
<p>This is a crucial step, as we need to make the data proper for the analysis. The first step is to change the column names, so it is easier to manipulate the dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="fu">library</span>(janitor)</span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a>qpcr <span class="ot">&lt;-</span> janitor<span class="sc">::</span><span class="fu">clean_names</span>(qpcr)</span>
<span id="cb9-4"><a href="#cb9-4"></a></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="fu">colnames</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "well"          "well_position" "omit"          "sample_name"  
[5] "target_name"   "ct"           </code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>We saw previously that the CT column is a character column. It should be a numeric instead. At least based on this qPCR machine, sometimes the CT will be set to <em>Undetermined</em>, so we actually need to replace it to an NA and then convert the column to numeric.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>qpcr <span class="ot">&lt;-</span> qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">ct =</span> <span class="fu">ifelse</span>(</span>
<span id="cb11-3"><a href="#cb11-3"></a>        ct <span class="sc">==</span> <span class="st">"Undetermined"</span>,</span>
<span id="cb11-4"><a href="#cb11-4"></a>        <span class="cn">NA</span>, </span>
<span id="cb11-5"><a href="#cb11-5"></a>        ct</span>
<span id="cb11-6"><a href="#cb11-6"></a>    )) <span class="sc">%&gt;%</span> </span>
<span id="cb11-7"><a href="#cb11-7"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">ct =</span> <span class="fu">as.numeric</span>(ct))</span>
<span id="cb11-8"><a href="#cb11-8"></a></span>
<span id="cb11-9"><a href="#cb11-9"></a>qpcr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 180 × 6
    well well_position omit  sample_name    target_name    ct
   &lt;dbl&gt; &lt;chr&gt;         &lt;lgl&gt; &lt;chr&gt;          &lt;chr&gt;       &lt;dbl&gt;
 1   195 I3            FALSE 1. eGFP N=1    CDH1         18.4
 2   196 I4            FALSE 1. eGFP N=1    CDH1         18.3
 3   219 J3            FALSE 1. eGFP N=1    CDH1         18.2
 4   245 K5            FALSE 10. ZEB2 N=3   CDH1         18.6
 5   246 K6            FALSE 10. ZEB2 N=3   CDH1         18.5
 6   269 L5            FALSE 10. ZEB2 N=3   CDH1         18.4
 7   293 M5            FALSE 11. TWIST1 N=3 CDH1         17.8
 8   294 M6            FALSE 11. TWIST1 N=3 CDH1         17.7
 9   317 N5            FALSE 11. TWIST1 N=3 CDH1         17.5
10   341 O5            FALSE 12. SNAI1 N=3  CDH1         17.9
# … with 170 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>Just to double check we can calculate a summary of the CT column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">summary</span>(qpcr<span class="sc">$</span>ct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
  14.90   17.61   26.27   27.12   36.52   43.78       2 </code></pre>
</div>
</div>
<p>There are two NA’s. We can actually check what are those.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>qpcr <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="fu">is.na</span>(ct))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 6
   well well_position omit  sample_name target_name    ct
  &lt;dbl&gt; &lt;chr&gt;         &lt;lgl&gt; &lt;chr&gt;       &lt;chr&gt;       &lt;dbl&gt;
1     7 A7            FALSE 1. eGFP N=1 ZEB2           NA
2    10 A10           FALSE 9. eGFP N=3 ZEB2           NA</code></pre>
</div>
</div>
</section>
<section id="formatting-sample-names" class="slide level2">
<h2>Formatting sample names</h2>
<p>The next crucial step is to have columns that gives us some information on the samples. This we can extract from sample_name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb17-2"><a href="#cb17-2"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(sample_name, target_name) <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a>    head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  sample_name  target_name
  &lt;chr&gt;        &lt;chr&gt;      
1 1. eGFP N=1  CDH1       
2 1. eGFP N=1  CDH1       
3 1. eGFP N=1  CDH1       
4 10. ZEB2 N=3 CDH1       
5 10. ZEB2 N=3 CDH1       
6 10. ZEB2 N=3 CDH1       </code></pre>
</div>
</div>
<p>All of the samples start with a number, followed by a dot and then their conditions. The first numbers correspond to the technical replicate, then it is following by the condition and the N=i corresponds to which biological replicate it is referring to.</p>
</section>
<section class="slide level2">

<p>We want to have the information for each sample and their technical replicates as well so when we summarize and visualize this can be taken into account.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb19-2"><a href="#cb19-2"></a></span>
<span id="cb19-3"><a href="#cb19-3"></a>qpcr <span class="ot">&lt;-</span> qpcr <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>    tidyr<span class="sc">::</span><span class="fu">separate</span>(</span>
<span id="cb19-5"><a href="#cb19-5"></a>        sample_name, </span>
<span id="cb19-6"><a href="#cb19-6"></a>        <span class="at">remove =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-7"><a href="#cb19-7"></a>        <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"sample_nb"</span>, <span class="st">"sample"</span>, <span class="st">"tech_replicate"</span>),</span>
<span id="cb19-8"><a href="#cb19-8"></a>        <span class="at">sep =</span> <span class="st">" "</span></span>
<span id="cb19-9"><a href="#cb19-9"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="summarizing" class="slide level2">
<h2>Summarizing</h2>
<p>The next thing is to summarize the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>summarised_qpcr <span class="ot">&lt;-</span> qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb20-2"><a href="#cb20-2"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(</span>
<span id="cb20-3"><a href="#cb20-3"></a>        sample_name, tech_replicate, sample, sample_nb, target_name</span>
<span id="cb20-4"><a href="#cb20-4"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb20-6"><a href="#cb20-6"></a>        <span class="at">mean =</span> <span class="fu">mean</span>(ct, <span class="at">na.rm =</span> <span class="cn">TRUE</span>),</span>
<span id="cb20-7"><a href="#cb20-7"></a>        <span class="at">sd =</span> <span class="fu">sd</span>(ct, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-8"><a href="#cb20-8"></a>    )</span>
<span id="cb20-9"><a href="#cb20-9"></a></span>
<span id="cb20-10"><a href="#cb20-10"></a>summarised_qpcr <span class="sc">%&gt;%</span> head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
# Groups:   sample_name, tech_replicate, sample, sample_nb [2]
  sample_name  tech_replicate sample sample_nb target_name  mean     sd
  &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt;
1 1. eGFP N=1  N=1            eGFP   1.        CDH1         18.3 0.118 
2 1. eGFP N=1  N=1            eGFP   1.        GAPDH        16.0 0.0589
3 1. eGFP N=1  N=1            eGFP   1.        SNAI1        35.9 2.84  
4 1. eGFP N=1  N=1            eGFP   1.        TWIST1       32.3 0.564 
5 1. eGFP N=1  N=1            eGFP   1.        ZEB2         39.3 2.31  
6 10. ZEB2 N=3 N=3            ZEB2   10.       CDH1         18.5 0.0566</code></pre>
</div>
</div>
</section>
<section id="plotting" class="slide level2">
<h2>Plotting</h2>
<p>We want to investigate if there is any outlier in these measurements, usually this can be picked up by the standard deviation of the samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(</span>
<span id="cb22-4"><a href="#cb22-4"></a>    summarised_qpcr, </span>
<span id="cb22-5"><a href="#cb22-5"></a>    <span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> mean, <span class="at">color =</span> target_name, <span class="at">shape =</span> tech_replicate)</span>
<span id="cb22-6"><a href="#cb22-6"></a>) <span class="sc">+</span> </span>
<span id="cb22-7"><a href="#cb22-7"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb22-8"><a href="#cb22-8"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Average CT values"</span>) <span class="sc">+</span></span>
<span id="cb22-9"><a href="#cb22-9"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-14-1.png" width="960" class="r-stretch"></section>
<section id="looking-for-outliers" class="slide level2">
<h2>Looking for outliers</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(</span>
<span id="cb23-2"><a href="#cb23-2"></a>    summarised_qpcr, </span>
<span id="cb23-3"><a href="#cb23-3"></a>    <span class="fu">aes</span>(</span>
<span id="cb23-4"><a href="#cb23-4"></a>        <span class="at">x =</span> mean, </span>
<span id="cb23-5"><a href="#cb23-5"></a>        <span class="at">y =</span> sd, </span>
<span id="cb23-6"><a href="#cb23-6"></a>        <span class="at">color =</span> target_name, </span>
<span id="cb23-7"><a href="#cb23-7"></a>        <span class="at">shape =</span> sample</span>
<span id="cb23-8"><a href="#cb23-8"></a>    )</span>
<span id="cb23-9"><a href="#cb23-9"></a>) <span class="sc">+</span> </span>
<span id="cb23-10"><a href="#cb23-10"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb23-11"><a href="#cb23-11"></a>    <span class="fu">labs</span>(</span>
<span id="cb23-12"><a href="#cb23-12"></a>        <span class="at">x =</span> <span class="st">"Average CT values"</span>,</span>
<span id="cb23-13"><a href="#cb23-13"></a>        <span class="at">y =</span> <span class="st">"SD of CT values"</span></span>
<span id="cb23-14"><a href="#cb23-14"></a>    ) <span class="sc">+</span></span>
<span id="cb23-15"><a href="#cb23-15"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-16-1.png" width="960" class="r-stretch"><p>Already from here we see that samples with high average CT values have high SD, which makes sense.</p>
</section>
<section id="normalizing" class="slide level2">
<h2>Normalizing</h2>
<p>Once we explored the data we can move to the normalization. Here we use the <span class="math inline">\(\Delta\)</span>CT procedure. We basically extract the geometric average of CT values from the housekeeping genes and then we can calculate log2 fold changes.</p>
</section>
<section class="slide level2">

<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>genes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"CDH1"</span>, <span class="st">"ZEB2"</span>, <span class="st">"SNAI1"</span>, <span class="st">"TWIST1"</span>)</span>
<span id="cb24-2"><a href="#cb24-2"></a>hk_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GAPDH"</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a></span>
<span id="cb24-4"><a href="#cb24-4"></a>normalized_qpcr <span class="ot">&lt;-</span> summarised_qpcr <span class="sc">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-6"><a href="#cb24-6"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb24-7"><a href="#cb24-7"></a>        <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"sample_name"</span>, <span class="st">"tech_replicate"</span>, <span class="st">"sample"</span>, <span class="st">"sample_nb"</span>),</span>
<span id="cb24-8"><a href="#cb24-8"></a>        <span class="at">values_from =</span> <span class="st">"mean"</span>,</span>
<span id="cb24-9"><a href="#cb24-9"></a>        <span class="at">names_from =</span> <span class="st">"target_name"</span></span>
<span id="cb24-10"><a href="#cb24-10"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb24-11"><a href="#cb24-11"></a>    <span class="co"># apply the operation in each row individually</span></span>
<span id="cb24-12"><a href="#cb24-12"></a>    dplyr<span class="sc">::</span><span class="fu">rowwise</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb24-13"><a href="#cb24-13"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">hk_geom =</span> <span class="fu">exp</span>(<span class="fu">mean</span>(<span class="fu">log</span>(<span class="sc">!!</span><span class="fu">sym</span>(hk_genes))))) <span class="sc">%&gt;%</span></span>
<span id="cb24-14"><a href="#cb24-14"></a>    <span class="co"># now we convert back to the long format but only for the </span></span>
<span id="cb24-15"><a href="#cb24-15"></a>    <span class="co"># genes of interest, so we can apply for each row the difference</span></span>
<span id="cb24-16"><a href="#cb24-16"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb24-17"><a href="#cb24-17"></a>        <span class="at">cols =</span> <span class="fu">all_of</span>(genes_of_interest),</span>
<span id="cb24-18"><a href="#cb24-18"></a>        <span class="at">names_to =</span> <span class="st">"target_name"</span>,</span>
<span id="cb24-19"><a href="#cb24-19"></a>        <span class="at">values_to =</span> <span class="st">"mean"</span></span>
<span id="cb24-20"><a href="#cb24-20"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb24-21"><a href="#cb24-21"></a>    <span class="fu">mutate</span>(<span class="at">dct =</span> mean <span class="sc">-</span> hk_geom)</span>
<span id="cb24-22"><a href="#cb24-22"></a>    </span>
<span id="cb24-23"><a href="#cb24-23"></a>normalized_qpcr <span class="sc">%&gt;%</span> head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  sample_name  tech_replicate sample sample_nb GAPDH hk_geom targe…¹  mean   dct
  &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 1. eGFP N=1  N=1            eGFP   1.         16.0    16.0 CDH1     18.3  2.33
2 1. eGFP N=1  N=1            eGFP   1.         16.0    16.0 ZEB2     39.3 23.3 
3 1. eGFP N=1  N=1            eGFP   1.         16.0    16.0 SNAI1    35.9 19.9 
4 1. eGFP N=1  N=1            eGFP   1.         16.0    16.0 TWIST1   32.3 16.4 
5 10. ZEB2 N=3 N=3            ZEB2   10.        16.0    16.0 CDH1     18.5  2.53
6 10. ZEB2 N=3 N=3            ZEB2   10.        16.0    16.0 ZEB2     25.7  9.72
# … with abbreviated variable name ¹​target_name</code></pre>
</div>
</div>
</section>
<section id="plotting-once-again-very-important" class="slide level2">
<h2>Plotting once again (very important!!!)</h2>
<h3 id="deltact"><span class="math inline">\(\Delta\)</span>CT</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>normalized_qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb26-2"><a href="#cb26-2"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(sample <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"ZEB2"</span>, <span class="st">"eGFP"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> dct, <span class="at">x =</span> sample, <span class="at">color =</span> target_name)) <span class="sc">+</span> </span>
<span id="cb26-4"><a href="#cb26-4"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb26-5"><a href="#cb26-5"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Condition"</span>, <span class="at">y =</span> <span class="fu">expression</span>(Delta<span class="sc">*</span><span class="st">"CT"</span>)) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-19-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<h3 id="deltact-stratified-by-target"><span class="math inline">\(\Delta\)</span>CT: stratified by target</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>normalized_qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb27-2"><a href="#cb27-2"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> dct, <span class="at">x =</span> sample, <span class="at">color =</span> target_name)) <span class="sc">+</span> </span>
<span id="cb27-3"><a href="#cb27-3"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb27-4"><a href="#cb27-4"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(</span>
<span id="cb27-5"><a href="#cb27-5"></a>        <span class="at">x =</span> <span class="st">"Condition"</span>, </span>
<span id="cb27-6"><a href="#cb27-6"></a>        <span class="at">y =</span> <span class="fu">expression</span>(Delta<span class="sc">*</span><span class="st">"CT"</span>),</span>
<span id="cb27-7"><a href="#cb27-7"></a>        <span class="at">title =</span> <span class="fu">expression</span>(Delta<span class="sc">*</span><span class="st">"CT by target"</span>)</span>
<span id="cb27-8"><a href="#cb27-8"></a>    ) <span class="sc">+</span></span>
<span id="cb27-9"><a href="#cb27-9"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>target_name) <span class="sc">+</span></span>
<span id="cb27-10"><a href="#cb27-10"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-21-1.png" width="960" class="r-stretch"></section>
<section id="checking-ct-values-and-normalized-data" class="slide level2">
<h2>Checking CT values and normalized data</h2>

<img data-src="slides_files/figure-revealjs/unnamed-chunk-22-1.png" width="960" class="r-stretch"></section>
<section id="statistical-testing" class="slide level2">
<h2>Statistical testing</h2>
<p>Lastly, we want to generate the statistical report. All the comparisons will be made against the <em>eGFP</em> sample. We will use the linear regression framework, which is similar to doing a t-test.</p>
<p>We want to do each comparison for each overexpression experiment and each target gene. The only comparisons we are doing is against the control, as this will answer our initial question.</p>
</section>
<section class="slide level2">

<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="fu">library</span>(purrr)</span>
<span id="cb28-3"><a href="#cb28-3"></a></span>
<span id="cb28-4"><a href="#cb28-4"></a>samples_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ZEB2"</span>, <span class="st">"TWIST1"</span>, <span class="st">"SNAI1"</span>)</span>
<span id="cb28-5"><a href="#cb28-5"></a>control_samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"eGFP"</span>, <span class="st">"eGFP"</span>, <span class="st">"eGFP"</span>)</span>
<span id="cb28-6"><a href="#cb28-6"></a></span>
<span id="cb28-7"><a href="#cb28-7"></a>stats_results <span class="ot">&lt;-</span> <span class="fu">mapply</span>(</span>
<span id="cb28-8"><a href="#cb28-8"></a>    <span class="cf">function</span>(condition_1, condition_2, normalized_qpcr){</span>
<span id="cb28-9"><a href="#cb28-9"></a>        normalized_qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb28-10"><a href="#cb28-10"></a>            dplyr<span class="sc">::</span><span class="fu">filter</span>(sample <span class="sc">%in%</span> <span class="fu">c</span>(condition_1, condition_2)) <span class="sc">%&gt;%</span></span>
<span id="cb28-11"><a href="#cb28-11"></a>            <span class="co"># Now we factor so the first level is control and the difference</span></span>
<span id="cb28-12"><a href="#cb28-12"></a>            <span class="co"># then represents treatment - control.</span></span>
<span id="cb28-13"><a href="#cb28-13"></a>            dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb28-14"><a href="#cb28-14"></a>                <span class="at">sample =</span> <span class="fu">factor</span>(sample, <span class="at">levels =</span> <span class="fu">c</span>(condition_2, condition_1))</span>
<span id="cb28-15"><a href="#cb28-15"></a>            ) <span class="sc">%&gt;%</span> </span>
<span id="cb28-16"><a href="#cb28-16"></a>            dplyr<span class="sc">::</span><span class="fu">group_by</span>(target_name) <span class="sc">%&gt;%</span></span>
<span id="cb28-17"><a href="#cb28-17"></a>            tidyr<span class="sc">::</span><span class="fu">nest</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-18"><a href="#cb28-18"></a>            <span class="co"># perform the linear regression and tidy up</span></span>
<span id="cb28-19"><a href="#cb28-19"></a>            dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb28-20"><a href="#cb28-20"></a>                <span class="at">results =</span> purrr<span class="sc">::</span><span class="fu">map</span>(</span>
<span id="cb28-21"><a href="#cb28-21"></a>                    purrr<span class="sc">::</span><span class="fu">map</span>(data, <span class="sc">~</span><span class="fu">lm</span>(dct <span class="sc">~</span> sample, <span class="at">data =</span> .)),</span>
<span id="cb28-22"><a href="#cb28-22"></a>                    broom<span class="sc">::</span>tidy</span>
<span id="cb28-23"><a href="#cb28-23"></a>                )</span>
<span id="cb28-24"><a href="#cb28-24"></a>            ) <span class="sc">%&gt;%</span></span>
<span id="cb28-25"><a href="#cb28-25"></a>            tidyr<span class="sc">::</span><span class="fu">unnest</span>(results)</span>
<span id="cb28-26"><a href="#cb28-26"></a>    },</span>
<span id="cb28-27"><a href="#cb28-27"></a>    samples_of_interest,</span>
<span id="cb28-28"><a href="#cb28-28"></a>    control_samples,</span>
<span id="cb28-29"><a href="#cb28-29"></a>    <span class="at">MoreArgs =</span> <span class="fu">list</span>(<span class="at">normalized_qpcr =</span> normalized_qpcr),</span>
<span id="cb28-30"><a href="#cb28-30"></a>    <span class="at">USE.NAMES =</span> <span class="cn">TRUE</span>,</span>
<span id="cb28-31"><a href="#cb28-31"></a>    <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span></span>
<span id="cb28-32"><a href="#cb28-32"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb28-33"><a href="#cb28-33"></a>    dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(<span class="at">.id =</span> <span class="st">"sample"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-34"><a href="#cb28-34"></a>    <span class="co"># remove the data column, not necessary to repeat the normalized qpcr df</span></span>
<span id="cb28-35"><a href="#cb28-35"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>tidyr<span class="sc">::</span><span class="fu">all_of</span>(<span class="st">"data"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb28-36"><a href="#cb28-36"></a>    janitor<span class="sc">::</span><span class="fu">clean_names</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-37"><a href="#cb28-37"></a>    <span class="co"># we tidy up the terms now</span></span>
<span id="cb28-38"><a href="#cb28-38"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">term =</span> <span class="fu">ifelse</span>(</span>
<span id="cb28-39"><a href="#cb28-39"></a>        term <span class="sc">==</span> <span class="st">"(Intercept)"</span>, </span>
<span id="cb28-40"><a href="#cb28-40"></a>        <span class="st">"ctrl_delta_ct"</span>,</span>
<span id="cb28-41"><a href="#cb28-41"></a>        <span class="st">"delta_delta_ct"</span></span>
<span id="cb28-42"><a href="#cb28-42"></a>    )) <span class="sc">%&gt;%</span></span>
<span id="cb28-43"><a href="#cb28-43"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb28-44"><a href="#cb28-44"></a>        <span class="at">id_cols =</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"target_name"</span>),</span>
<span id="cb28-45"><a href="#cb28-45"></a>        <span class="at">names_from =</span> <span class="st">"term"</span>,</span>
<span id="cb28-46"><a href="#cb28-46"></a>        <span class="at">values_from =</span> <span class="fu">c</span>(<span class="st">"estimate"</span>, <span class="st">"std_error"</span>, <span class="st">"statistic"</span>, <span class="st">"p_value"</span>)</span>
<span id="cb28-47"><a href="#cb28-47"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb28-48"><a href="#cb28-48"></a>    <span class="co"># if necessary, do adjustment of p-values</span></span>
<span id="cb28-49"><a href="#cb28-49"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb28-50"><a href="#cb28-50"></a>        <span class="at">padj_delta_delta_ct =</span> <span class="fu">p.adjust</span>(p_value_delta_delta_ct),</span>
<span id="cb28-51"><a href="#cb28-51"></a>        <span class="at">fc =</span> <span class="dv">2</span><span class="sc">^</span>(<span class="sc">-</span>estimate_delta_delta_ct),</span>
<span id="cb28-52"><a href="#cb28-52"></a>        <span class="at">logfc =</span> <span class="sc">-</span>estimate_delta_delta_ct</span>
<span id="cb28-53"><a href="#cb28-53"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb28-54"><a href="#cb28-54"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(padj_delta_delta_ct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">

<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb29-2"><a href="#cb29-2"></a>readr<span class="sc">::</span><span class="fu">write_csv</span>(stats_results, <span class="st">"../results/stats_results.csv"</span>)</span>
<span id="cb29-3"><a href="#cb29-3"></a></span>
<span id="cb29-4"><a href="#cb29-4"></a>stats_results <span class="sc">%&gt;%</span> </span>
<span id="cb29-5"><a href="#cb29-5"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(</span>
<span id="cb29-6"><a href="#cb29-6"></a>        sample, target_name, logfc, </span>
<span id="cb29-7"><a href="#cb29-7"></a>        fc, padj_delta_delta_ct, </span>
<span id="cb29-8"><a href="#cb29-8"></a>        p_value_delta_delta_ct</span>
<span id="cb29-9"><a href="#cb29-9"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb29-10"><a href="#cb29-10"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(padj_delta_delta_ct) <span class="sc">%&gt;%</span></span>
<span id="cb29-11"><a href="#cb29-11"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">pvalue =</span> p_value_delta_delta_ct) <span class="sc">%&gt;%</span></span>
<span id="cb29-12"><a href="#cb29-12"></a>    head</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
# Groups:   target_name [4]
  sample target_name  logfc       fc padj_delta_delta_ct   pvalue
  &lt;chr&gt;  &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;               &lt;dbl&gt;    &lt;dbl&gt;
1 TWIST1 TWIST1       8.85   462.               0.000387 0.000129
2 ZEB2   ZEB2        13.0   8072.               0.00244  0.000814
3 SNAI1  SNAI1        3.01     8.07             0.149    0.0497  
4 ZEB2   CDH1        -0.843    0.557            0.654    0.218   
5 ZEB2   SNAI1       -0.735    0.601            0.881    0.495   
6 TWIST1 SNAI1       -0.605    0.658            0.881    0.440   </code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>Only two genes end up being differentially expressed when compared to control, with high fold change, but here we always need to take into account the cycle threshold for all the conditions. Having a CT close to 40 means the gene is either barely detectable or not detectable at all, so the fold change can be misleading.</p>
</section></section>
<section>
<section id="visualizing-the-results" class="title-slide slide level1 center">
<h1>Visualizing the results</h1>
<p>In the next few slides I will show different ways of visualizing the results, they are highly inspired in RNA-plots to visualize the results and check for potential outliers.</p>
<p>We start by showing the code and the next slide shows the plot.</p>
</section>
<section id="logfc-and-p-values" class="slide level2">
<h2>logFC and p-values</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>stats_results <span class="sc">%&gt;%</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(</span>
<span id="cb31-3"><a href="#cb31-3"></a>        <span class="fu">aes</span>(</span>
<span id="cb31-4"><a href="#cb31-4"></a>            <span class="at">x =</span> estimate_ctrl_delta_ct, </span>
<span id="cb31-5"><a href="#cb31-5"></a>            <span class="at">y =</span> logfc,</span>
<span id="cb31-6"><a href="#cb31-6"></a>            <span class="at">shape =</span> sample,</span>
<span id="cb31-7"><a href="#cb31-7"></a>            <span class="at">color =</span> p_value_delta_delta_ct</span>
<span id="cb31-8"><a href="#cb31-8"></a>        )</span>
<span id="cb31-9"><a href="#cb31-9"></a>    ) <span class="sc">+</span></span>
<span id="cb31-10"><a href="#cb31-10"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb31-11"><a href="#cb31-11"></a>    ggplot2<span class="sc">::</span><span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb31-12"><a href="#cb31-12"></a>    <span class="co"># this wraps around the target genes from the qPCR data. </span></span>
<span id="cb31-13"><a href="#cb31-13"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>target_name) <span class="sc">+</span> </span>
<span id="cb31-14"><a href="#cb31-14"></a>    <span class="co"># base_size indicates the overall size of the fonts in the plot</span></span>
<span id="cb31-15"><a href="#cb31-15"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-25-1.png" width="960" class="r-stretch"></section>
<section id="logfc-and-barplots" class="slide level2">
<h2>logFC and barplots</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>normalized_qpcr <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(sample, target_name) <span class="sc">%&gt;%</span></span>
<span id="cb32-3"><a href="#cb32-3"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">mean_hk =</span> <span class="fu">mean</span>(hk_geom), <span class="at">.groups =</span> <span class="st">"keep"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>    <span class="co"># concatenate two dataframes based on the values in the columns </span></span>
<span id="cb32-5"><a href="#cb32-5"></a>    <span class="co"># sample and target_name, useful to join information that are in</span></span>
<span id="cb32-6"><a href="#cb32-6"></a>    <span class="co"># different tables. right_join includes all the rows of the </span></span>
<span id="cb32-7"><a href="#cb32-7"></a>    <span class="co"># dataframe stats_results. There are other functions to perform similar</span></span>
<span id="cb32-8"><a href="#cb32-8"></a>    <span class="co"># actions, such as inner_join, left_join and full_join</span></span>
<span id="cb32-9"><a href="#cb32-9"></a>    dplyr<span class="sc">::</span><span class="fu">right_join</span>(., stats_results, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"target_name"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb32-10"><a href="#cb32-10"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">is_sig =</span> <span class="fu">ifelse</span>(</span>
<span id="cb32-11"><a href="#cb32-11"></a>        padj_delta_delta_ct <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb32-12"><a href="#cb32-12"></a>        <span class="st">"yes"</span>,</span>
<span id="cb32-13"><a href="#cb32-13"></a>        <span class="st">"no"</span></span>
<span id="cb32-14"><a href="#cb32-14"></a>    )) <span class="sc">%&gt;%</span></span>
<span id="cb32-15"><a href="#cb32-15"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(</span>
<span id="cb32-16"><a href="#cb32-16"></a>        <span class="fu">aes</span>(</span>
<span id="cb32-17"><a href="#cb32-17"></a>            <span class="at">x =</span> estimate_ctrl_delta_ct, </span>
<span id="cb32-18"><a href="#cb32-18"></a>            <span class="at">y =</span> logfc,</span>
<span id="cb32-19"><a href="#cb32-19"></a>            <span class="at">color =</span> mean_hk,</span>
<span id="cb32-20"><a href="#cb32-20"></a>            <span class="at">ymin =</span> logfc <span class="sc">-</span> std_error_delta_delta_ct,</span>
<span id="cb32-21"><a href="#cb32-21"></a>            <span class="at">ymax =</span> logfc <span class="sc">+</span> std_error_delta_delta_ct</span>
<span id="cb32-22"><a href="#cb32-22"></a>        )</span>
<span id="cb32-23"><a href="#cb32-23"></a>    ) <span class="sc">+</span></span>
<span id="cb32-24"><a href="#cb32-24"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_col</span>(<span class="at">size =</span> <span class="dv">20</span>) <span class="sc">+</span> </span>
<span id="cb32-25"><a href="#cb32-25"></a>    <span class="co"># change the color to the viridis scale, which is easier to read from</span></span>
<span id="cb32-26"><a href="#cb32-26"></a>    <span class="co"># and if printed in gray-scale one can still see the differences</span></span>
<span id="cb32-27"><a href="#cb32-27"></a>    ggplot2<span class="sc">::</span><span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb32-28"><a href="#cb32-28"></a>    <span class="co"># add the line in the barplots</span></span>
<span id="cb32-29"><a href="#cb32-29"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_crossbar</span>(<span class="at">fatten =</span> <span class="dv">4</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb32-30"><a href="#cb32-30"></a>    <span class="co"># wraps the plot in two variables, samples are represented</span></span>
<span id="cb32-31"><a href="#cb32-31"></a>    <span class="co"># in the rows and the target genes are in the columns</span></span>
<span id="cb32-32"><a href="#cb32-32"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_grid</span>(sample <span class="sc">~</span> target_name, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span> </span>
<span id="cb32-33"><a href="#cb32-33"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb32-34"><a href="#cb32-34"></a>    <span class="co"># change your axis names here with the labs function</span></span>
<span id="cb32-35"><a href="#cb32-35"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(</span>
<span id="cb32-36"><a href="#cb32-36"></a>        <span class="at">color =</span> <span class="st">"Mean of</span><span class="sc">\n</span><span class="st">HK genes"</span>,</span>
<span id="cb32-37"><a href="#cb32-37"></a>        <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">"log"</span>[<span class="dv">2</span>]<span class="sc">*</span><span class="st">"FC"</span><span class="sc">^</span><span class="dv">1</span>),</span>
<span id="cb32-38"><a href="#cb32-38"></a>        <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Average "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT in CTRL samples"</span>),</span>
<span id="cb32-39"><a href="#cb32-39"></a>        <span class="at">caption =</span> <span class="fu">expression</span>(<span class="st">" "</span><span class="sc">^</span><span class="dv">1</span><span class="sc">*</span><span class="st">"Difference between "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of control and other condition"</span>)</span>
<span id="cb32-40"><a href="#cb32-40"></a>    ) <span class="sc">+</span></span>
<span id="cb32-41"><a href="#cb32-41"></a>    <span class="co"># we flip the x axis values so it is easier to read as in this way </span></span>
<span id="cb32-42"><a href="#cb32-42"></a>    <span class="co"># they are spaced</span></span>
<span id="cb32-43"><a href="#cb32-43"></a>    ggplot2<span class="sc">::</span><span class="fu">theme</span>(</span>
<span id="cb32-44"><a href="#cb32-44"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(</span>
<span id="cb32-45"><a href="#cb32-45"></a>            <span class="at">angle =</span> <span class="dv">45</span>, </span>
<span id="cb32-46"><a href="#cb32-46"></a>            <span class="at">vjust =</span> <span class="dv">1</span>, </span>
<span id="cb32-47"><a href="#cb32-47"></a>            <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb32-48"><a href="#cb32-48"></a>        )</span>
<span id="cb32-49"><a href="#cb32-49"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-26-1.png" width="960" class="r-stretch"></section>
<section id="logfc-hk-genes-and-statistical-significance" class="slide level2">
<h2>logFC, HK genes and statistical significance</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>all_hk_genes <span class="ot">&lt;-</span> <span class="fu">paste</span>(hk_genes, <span class="at">collapse =</span> <span class="st">", "</span>)</span>
<span id="cb33-2"><a href="#cb33-2"></a></span>
<span id="cb33-3"><a href="#cb33-3"></a>normalized_qpcr <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(sample, target_name) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">mean_hk =</span> <span class="fu">mean</span>(hk_geom), <span class="at">.groups =</span> <span class="st">"keep"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-6"><a href="#cb33-6"></a>    dplyr<span class="sc">::</span><span class="fu">right_join</span>(., stats_results, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"target_name"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb33-7"><a href="#cb33-7"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb33-8"><a href="#cb33-8"></a>        <span class="at">is_sig =</span> <span class="fu">ifelse</span>(</span>
<span id="cb33-9"><a href="#cb33-9"></a>            padj_delta_delta_ct <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb33-10"><a href="#cb33-10"></a>            <span class="st">"yes"</span>,</span>
<span id="cb33-11"><a href="#cb33-11"></a>            <span class="st">"no"</span></span>
<span id="cb33-12"><a href="#cb33-12"></a>        ), </span>
<span id="cb33-13"><a href="#cb33-13"></a>        <span class="at">is_sig =</span> <span class="fu">factor</span>(is_sig, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"yes"</span>, <span class="st">"no"</span>))</span>
<span id="cb33-14"><a href="#cb33-14"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb33-15"><a href="#cb33-15"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(</span>
<span id="cb33-16"><a href="#cb33-16"></a>        <span class="fu">aes</span>(</span>
<span id="cb33-17"><a href="#cb33-17"></a>            <span class="at">x =</span> estimate_ctrl_delta_ct, </span>
<span id="cb33-18"><a href="#cb33-18"></a>            <span class="at">y =</span> logfc,</span>
<span id="cb33-19"><a href="#cb33-19"></a>            <span class="at">shape =</span> is_sig,</span>
<span id="cb33-20"><a href="#cb33-20"></a>            <span class="at">color =</span> mean_hk</span>
<span id="cb33-21"><a href="#cb33-21"></a>        )</span>
<span id="cb33-22"><a href="#cb33-22"></a>    ) <span class="sc">+</span></span>
<span id="cb33-23"><a href="#cb33-23"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb33-24"><a href="#cb33-24"></a>    ggplot2<span class="sc">::</span><span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb33-25"><a href="#cb33-25"></a>    <span class="co"># Here we leave the scale free in the x-axis, you can change the </span></span>
<span id="cb33-26"><a href="#cb33-26"></a>    <span class="co"># y-scale for each subindividual plot as well. Options are </span></span>
<span id="cb33-27"><a href="#cb33-27"></a>    <span class="co"># free_y, free and free_x.</span></span>
<span id="cb33-28"><a href="#cb33-28"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_grid</span>(sample <span class="sc">~</span> target_name, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span> </span>
<span id="cb33-29"><a href="#cb33-29"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb33-30"><a href="#cb33-30"></a>        ggplot2<span class="sc">::</span><span class="fu">labs</span>(</span>
<span id="cb33-31"><a href="#cb33-31"></a>        <span class="at">color =</span> <span class="fu">expression</span>(<span class="st">"Mean of</span><span class="sc">\n</span><span class="st">HK genes"</span><span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb33-32"><a href="#cb33-32"></a>        <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">"log"</span>[<span class="dv">2</span>]<span class="sc">*</span><span class="st">"FC"</span><span class="sc">^</span><span class="dv">1</span>),</span>
<span id="cb33-33"><a href="#cb33-33"></a>        <span class="at">x =</span> <span class="fu">expression</span>(<span class="st">"Average "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT in CTRL samples"</span>),</span>
<span id="cb33-34"><a href="#cb33-34"></a>        <span class="at">caption =</span> <span class="fu">expression</span>(<span class="st">" "</span><span class="sc">^</span><span class="dv">1</span><span class="sc">*</span><span class="st">"Difference between "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of control and other condition."</span>),</span>
<span id="cb33-35"><a href="#cb33-35"></a>        <span class="at">shape =</span> <span class="st">"padj &lt; 0.05?"</span></span>
<span id="cb33-36"><a href="#cb33-36"></a>    ) <span class="sc">+</span></span>
<span id="cb33-37"><a href="#cb33-37"></a>    ggplot2<span class="sc">::</span><span class="fu">theme</span>(</span>
<span id="cb33-38"><a href="#cb33-38"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(</span>
<span id="cb33-39"><a href="#cb33-39"></a>            <span class="at">angle =</span> <span class="dv">45</span>, </span>
<span id="cb33-40"><a href="#cb33-40"></a>            <span class="at">vjust =</span> <span class="dv">1</span>, </span>
<span id="cb33-41"><a href="#cb33-41"></a>            <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb33-42"><a href="#cb33-42"></a>        )</span>
<span id="cb33-43"><a href="#cb33-43"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-27-1.png" width="960" class="r-stretch"></section>
<section id="logfc-plot" class="slide level2">
<h2>logFC plot</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="co"># we use this library to add the p-values in the plot</span></span>
<span id="cb34-2"><a href="#cb34-2"></a><span class="fu">library</span>(ggprism)</span>
<span id="cb34-3"><a href="#cb34-3"></a></span>
<span id="cb34-4"><a href="#cb34-4"></a>caption_exp <span class="ot">&lt;-</span> <span class="fu">expression</span>(<span class="st">" "</span><span class="sc">^</span><span class="dv">1</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT difference: "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of control samples - "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of individual sample"</span>)</span>
<span id="cb34-5"><a href="#cb34-5"></a></span>
<span id="cb34-6"><a href="#cb34-6"></a><span class="co"># for simple comparisons we can use the ggprism package to add the p-values,</span></span>
<span id="cb34-7"><a href="#cb34-7"></a><span class="co"># if you have more conditions and comparisons in the same group you need </span></span>
<span id="cb34-8"><a href="#cb34-8"></a><span class="co"># to tinker the dataframe below to get the right y position in the plot</span></span>
<span id="cb34-9"><a href="#cb34-9"></a><span class="co"># for the bars.</span></span>
<span id="cb34-10"><a href="#cb34-10"></a>pvals_plot <span class="ot">&lt;-</span> stats_results <span class="sc">%&gt;%</span></span>
<span id="cb34-11"><a href="#cb34-11"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(padj_delta_delta_ct <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-12"><a href="#cb34-12"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb34-13"><a href="#cb34-13"></a>        <span class="at">group1 =</span> <span class="st">"eGFP"</span>,</span>
<span id="cb34-14"><a href="#cb34-14"></a>        <span class="at">group2 =</span> sample,</span>
<span id="cb34-15"><a href="#cb34-15"></a>        <span class="at">y.position =</span> logfc <span class="sc">+</span> std_error_delta_delta_ct,</span>
<span id="cb34-16"><a href="#cb34-16"></a>        <span class="at">label =</span> p_value_delta_delta_ct,</span>
<span id="cb34-17"><a href="#cb34-17"></a>        <span class="at">label =</span> <span class="fu">format</span>(label, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb34-18"><a href="#cb34-18"></a>    )</span>
<span id="cb34-19"><a href="#cb34-19"></a>    </span>
<span id="cb34-20"><a href="#cb34-20"></a></span>
<span id="cb34-21"><a href="#cb34-21"></a>normalized_qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb34-22"><a href="#cb34-22"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(sample <span class="sc">==</span> <span class="st">"eGFP"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-23"><a href="#cb34-23"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(target_name) <span class="sc">%&gt;%</span></span>
<span id="cb34-24"><a href="#cb34-24"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">avg_ctrl_dct =</span> <span class="fu">mean</span>(dct)) <span class="sc">%&gt;%</span></span>
<span id="cb34-25"><a href="#cb34-25"></a>    dplyr<span class="sc">::</span><span class="fu">right_join</span>(., normalized_qpcr, <span class="at">by =</span> <span class="st">"target_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-26"><a href="#cb34-26"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">logfc =</span> avg_ctrl_dct <span class="sc">-</span> dct) <span class="sc">%&gt;%</span></span>
<span id="cb34-27"><a href="#cb34-27"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> logfc, <span class="at">color =</span> sample)) <span class="sc">+</span></span>
<span id="cb34-28"><a href="#cb34-28"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span> </span>
<span id="cb34-29"><a href="#cb34-29"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb34-30"><a href="#cb34-30"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_wrap</span>(<span class="sc">~</span> target_name, <span class="at">scales =</span> <span class="st">"free_y"</span>, <span class="at">shrink =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb34-31"><a href="#cb34-31"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(</span>
<span id="cb34-32"><a href="#cb34-32"></a>        <span class="at">x =</span> <span class="st">"Samples"</span>,</span>
<span id="cb34-33"><a href="#cb34-33"></a>        <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">"log"</span>[<span class="dv">2</span>]<span class="sc">*</span><span class="st">"(Fold Change)"</span><span class="sc">^</span><span class="dv">1</span>),</span>
<span id="cb34-34"><a href="#cb34-34"></a>        <span class="at">caption =</span> caption_exp,</span>
<span id="cb34-35"><a href="#cb34-35"></a>        <span class="at">title =</span> <span class="st">"Overexpression of SNAI1, TWIST1 and ZEB2"</span>,</span>
<span id="cb34-36"><a href="#cb34-36"></a>        <span class="at">subtitle =</span> <span class="st">"p-values shown for samples with adjusted p-value &lt; 0.05"</span></span>
<span id="cb34-37"><a href="#cb34-37"></a>    ) <span class="sc">+</span></span>
<span id="cb34-38"><a href="#cb34-38"></a>    ggplot2<span class="sc">::</span><span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb34-39"><a href="#cb34-39"></a>    ggplot2<span class="sc">::</span><span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb34-40"><a href="#cb34-40"></a>    ggprism<span class="sc">::</span><span class="fu">add_pvalue</span>(</span>
<span id="cb34-41"><a href="#cb34-41"></a>        pvals_plot, </span>
<span id="cb34-42"><a href="#cb34-42"></a>        <span class="at">color =</span> <span class="st">"black"</span>, </span>
<span id="cb34-43"><a href="#cb34-43"></a>        <span class="at">tip.length =</span> <span class="dv">0</span>, </span>
<span id="cb34-44"><a href="#cb34-44"></a>        <span class="at">label.size =</span> <span class="dv">3</span></span>
<span id="cb34-45"><a href="#cb34-45"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-28-1.png" width="864" class="r-stretch"></section>
<section id="prism-theme" class="slide level2">
<h2>Prism theme</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb35" data-code-line-numbers="|40"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co"># we use this library to add the p-values in the plot</span></span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="fu">library</span>(ggprism)</span>
<span id="cb35-3"><a href="#cb35-3"></a></span>
<span id="cb35-4"><a href="#cb35-4"></a>caption_exp <span class="ot">&lt;-</span> <span class="fu">expression</span>(<span class="st">" "</span><span class="sc">^</span><span class="dv">1</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT difference: "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of control samples - "</span><span class="sc">*</span>Delta<span class="sc">*</span><span class="st">"CT of individual sample"</span>)</span>
<span id="cb35-5"><a href="#cb35-5"></a></span>
<span id="cb35-6"><a href="#cb35-6"></a><span class="co"># for simple comparisons we can use the ggprism package to add the p-values,</span></span>
<span id="cb35-7"><a href="#cb35-7"></a><span class="co"># if you have more conditions and comparisons in the same group you need </span></span>
<span id="cb35-8"><a href="#cb35-8"></a><span class="co"># to tinker the dataframe below to get the right y position in the plot</span></span>
<span id="cb35-9"><a href="#cb35-9"></a><span class="co"># for the bars.</span></span>
<span id="cb35-10"><a href="#cb35-10"></a>pvals_plot <span class="ot">&lt;-</span> stats_results <span class="sc">%&gt;%</span></span>
<span id="cb35-11"><a href="#cb35-11"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(padj_delta_delta_ct <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-12"><a href="#cb35-12"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb35-13"><a href="#cb35-13"></a>        <span class="at">group1 =</span> <span class="st">"eGFP"</span>,</span>
<span id="cb35-14"><a href="#cb35-14"></a>        <span class="at">group2 =</span> sample,</span>
<span id="cb35-15"><a href="#cb35-15"></a>        <span class="at">y.position =</span> logfc <span class="sc">+</span> <span class="dv">2</span><span class="sc">*</span>std_error_delta_delta_ct,</span>
<span id="cb35-16"><a href="#cb35-16"></a>        <span class="at">label =</span> p_value_delta_delta_ct,</span>
<span id="cb35-17"><a href="#cb35-17"></a>        <span class="at">label =</span> <span class="fu">format</span>(label, <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb35-18"><a href="#cb35-18"></a>    )</span>
<span id="cb35-19"><a href="#cb35-19"></a>    </span>
<span id="cb35-20"><a href="#cb35-20"></a></span>
<span id="cb35-21"><a href="#cb35-21"></a>normalized_qpcr <span class="sc">%&gt;%</span> </span>
<span id="cb35-22"><a href="#cb35-22"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(sample <span class="sc">==</span> <span class="st">"eGFP"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-23"><a href="#cb35-23"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(target_name) <span class="sc">%&gt;%</span></span>
<span id="cb35-24"><a href="#cb35-24"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">avg_ctrl_dct =</span> <span class="fu">mean</span>(dct)) <span class="sc">%&gt;%</span></span>
<span id="cb35-25"><a href="#cb35-25"></a>    dplyr<span class="sc">::</span><span class="fu">right_join</span>(., normalized_qpcr, <span class="at">by =</span> <span class="st">"target_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-26"><a href="#cb35-26"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">logfc =</span> avg_ctrl_dct <span class="sc">-</span> dct) <span class="sc">%&gt;%</span></span>
<span id="cb35-27"><a href="#cb35-27"></a>    ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sample, <span class="at">y =</span> logfc, <span class="at">color =</span> sample)) <span class="sc">+</span></span>
<span id="cb35-28"><a href="#cb35-28"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span> </span>
<span id="cb35-29"><a href="#cb35-29"></a>    ggplot2<span class="sc">::</span><span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb35-30"><a href="#cb35-30"></a>    ggplot2<span class="sc">::</span><span class="fu">facet_wrap</span>(<span class="sc">~</span> target_name, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb35-31"><a href="#cb35-31"></a>    ggplot2<span class="sc">::</span><span class="fu">labs</span>(</span>
<span id="cb35-32"><a href="#cb35-32"></a>        <span class="at">x =</span> <span class="st">"Samples"</span>,</span>
<span id="cb35-33"><a href="#cb35-33"></a>        <span class="at">y =</span> <span class="fu">expression</span>(<span class="st">"log"</span>[<span class="dv">2</span>]<span class="sc">*</span><span class="st">"(Fold Change)"</span><span class="sc">^</span><span class="dv">1</span>),</span>
<span id="cb35-34"><a href="#cb35-34"></a>        <span class="at">caption =</span> caption_exp,</span>
<span id="cb35-35"><a href="#cb35-35"></a>        <span class="at">title =</span> <span class="st">"Overexpression of SNAI1, TWIST1 and ZEB2"</span>,</span>
<span id="cb35-36"><a href="#cb35-36"></a>        <span class="at">subtitle =</span> <span class="st">"p-values shown for samples with adjusted p-value &lt; 0.05"</span></span>
<span id="cb35-37"><a href="#cb35-37"></a>    ) <span class="sc">+</span></span>
<span id="cb35-38"><a href="#cb35-38"></a>    ggplot2<span class="sc">::</span><span class="fu">scale_color_viridis_d</span>() <span class="sc">+</span> </span>
<span id="cb35-39"><a href="#cb35-39"></a>    <span class="co"># use the prism theme</span></span>
<span id="cb35-40"><a href="#cb35-40"></a>    ggprism<span class="sc">::</span><span class="fu">theme_prism</span>() <span class="sc">+</span></span>
<span id="cb35-41"><a href="#cb35-41"></a>    ggprism<span class="sc">::</span><span class="fu">add_pvalue</span>(</span>
<span id="cb35-42"><a href="#cb35-42"></a>        pvals_plot, </span>
<span id="cb35-43"><a href="#cb35-43"></a>        <span class="at">color =</span> <span class="st">"black"</span>,</span>
<span id="cb35-44"><a href="#cb35-44"></a>        <span class="at">tip.length =</span> <span class="dv">0</span>, </span>
<span id="cb35-45"><a href="#cb35-45"></a>        <span class="at">label.size =</span> <span class="dv">3</span></span>
<span id="cb35-46"><a href="#cb35-46"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">


<img data-src="slides_files/figure-revealjs/unnamed-chunk-29-1.png" width="864" class="r-stretch"></section></section>
<section>
<section id="generating-a-report" class="title-slide slide level1 center">
<h1>Generating a report</h1>

</section>
<section id="render-render-render" class="slide level2">
<h2>Render, Render, Render…</h2>
<p>We can now generate a report with the results of the analysis! For this, just click in the button “Render” in your Rstudio, this will generate a html or pdf file in your folder (just make sure to move your report to the results folder afterwards!).</p>

<img src="images/render_button.png" alt="Render button on Rstudio" class="r-stretch"><div class="footer footer-default">

</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        setTimeout(function() {
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const cites = ref.parentNode.getAttribute('data-cites').split(' ');
        tippyHover(ref, function() {
          var popup = window.document.createElement('div');
          cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    });
    </script>
    

</body></html>